modules/find.xq

```diff
@@ -22,47 +25,149 @@ import module namespace versions="http://exist-db.org/apps/public-repo/versions"
 declare namespace request="http://exist-db.org/xquery/request";
 declare namespace response="http://exist-db.org/xquery/response";
 
-let $abbrev := request:get-parameter("abbrev", ())
-let $name := request:get-parameter("name", ())
-let $exist-version-semver := request:get-parameter("processor", $config:default-exist-version)
-let $versions := request:get-parameter("version", ())
-let $semver := request:get-parameter("semver", ())
-let $semver-min := request:get-parameter("semver-min", ())
-let $semver-max := request:get-parameter("semver-max", ())
-let $zip := request:get-parameter("zip", ())
-let $info := request:get-parameter("info", ())
-let $app-root-absolute-url := request:get-parameter("app-root-absolute-url", ())
-
-let $packages :=
-    if ($name) then
+(: TODO shouldn't we get $abs-public from $config? - joewiz :)
+declare variable $app-root-absolute-url := request:get-parameter("app-root-absolute-url", ());
+declare variable $abs-public := $app-root-absolute-url || "/public/";
+
+declare variable $exist-version-semver := request:get-parameter("processor", $config:default-exist-version);
+declare variable $abbrev := request:get-parameter("abbrev", ());
+declare variable $name := request:get-parameter("name", ());
+declare variable $versions := request:get-parameter("version", ());
+declare variable $semver := request:get-parameter("semver", ());
+declare variable $semver-min := request:get-parameter("semver-min", ());
+declare variable $semver-max := request:get-parameter("semver-max", ());
+
+declare variable $zip := request:get-parameter("zip", ());
+declare variable $info := request:get-parameter("info", ());
+
+declare variable $versions-or-version-range :=
+        exists($versions)
+        or exists($semver)
+        or exists($semver-min)
+        or exists($semver-max)
+;
+
+(:
+ input: "text/html, application/xhtml+xml, application/xml;q=0.9, image/webp, */*;q=0.8"
+ output: ("text/html", "application/json", "application/xml", "image/webp", "*/*")
+:)
+declare function local:parse-accept-header($accept as xs:string) as xs:string* {
+    tokenize($accept, ",")
+        ! normalize-space() (: trim value :)
+        ! tokenize(., ";")[1] (: drop q :)
+};
+
+declare function local:prefers-json($mime-types as xs:string*) as xs:boolean {
+    let $json-index := index-of($mime-types, "application/json")
+    let $xml-index := index-of($mime-types, "application/xml")
+    return
+        exists($json-index) and (empty($xml-index) or $xml-index > $json-index)
```

Comment:
out of curiosity why would the condition after `or` ever get triggered?

Reply:
There is a somewhat hypothetical scenario that a client wants XML but also accepts JSON.

`Accept: application/xml; application/json`

Reply:
The _or_ condition checks the order of both mime-types to ensure that JSON is the first one.

This implementation does not take into account the q-values.

Reply:
But isn't it the q value that should determine preference, and not the actual ordering.

Reply:
> There is a somewhat hypothetical scenario that a client wants XML but also accepts JSON.

Is it hypothetical... Doesn't every web browser a d/or Ajax call do exactly that?

Reply:
Yes, if a q value is set it should take precedence. I dropped it because there is no client yet that even accepts JSON. So this is a known limitation at this point:

If JSON is accepted _and_ listed first we assume that the client does prefers JSON, otherwise it is just not listed at all (this is the case for all known versions of exist-db. Do you happen to know of a client where this would not suffice?

Reply:
I did in fact not try to `fetch()` yet. I might have to overthink this decision based on its standard accept header value.

Reply:
In all browser I tested `fetch` will default to `Accept: */*` 
This will return XML as it did before. Only by setting a header explicitly you can now control in what format the data is returned. So while - in theory - one can craft an accept header that will return XML instead of JSON and vice versa this is very unlikely to ever happen.